File: hw3/hw3/Driver.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 *
 * @author reubentanner
 */
public class Driver {

    private static FileAccessMonitor monitor;
    private static final int MAX_THREADS = 16;
    
    
    public static void main(String[] args) throws InterruptedException 
    {
        if (args.length != 1)
        {
            System.err.println("Usage: java Driver <n> ");
            System.exit(1);
        }
        
        int max = Integer.parseInt(args[0]);
        monitor = new FileAccessMonitor(max);
        Random rand = new Random(max);
        
        List<Thread> threads = new ArrayList<>();
        for (int i = 0; i < MAX_THREADS; ++i)
        {
           Thread fileAccessThread = new FileAccessThread(rand.nextInt(max), monitor);
           threads.add(fileAccessThread);
           fileAccessThread.start();
        }
        
        for (Thread thread : threads)
        {
            thread.join();
        }
    }
}

File: hw3/hw3/driver.c
//
//  driver.c
//  hw3
//
//  Created by Reuben Tanner on 11/3/14.
//  Copyright (c) 2014 Reuben Tanner. All rights reserved.
//

#include "file_acess_monitor.h"
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>

#define MAX_THREADS 16

void* operate_on_file(void * arg);
static void set_thread_name(int id);

int n;
monitor_ptr monitor;

int main(int argc, char** argv)
{
    if (argc < 2)
    {
        fprintf(stderr, "usage: ./driver <n>\n");
        exit(EXIT_FAILURE);
    }
    
    n = atoi(argv[1]);
    pthread_t threads[MAX_THREADS];
    
    monitor = create_monitor(n);
   
    int i;
    for (i = 0; i < MAX_THREADS; ++i)
    {
        pthread_create(&(threads[i]), NULL, operate_on_file, (void*)(rand() % monitor->max));
    }
    for (i = 0; i < MAX_THREADS; ++i)
    {
        pthread_join(threads[i], NULL);
    }
    
}

void* operate_on_file(void * arg)
{
    set_thread_name((int)arg);
    access_file(monitor);
    sleep(1);
    release_file(monitor);
    
    return NULL;
}

static void set_thread_name(int id)
{
    char name[16];
    sprintf(name, "%d", id);
    pthread_setname_np(name);
}